crypto.go

```diff
@@ -10,29 +10,41 @@ import (
 	"golang.org/x/crypto/nacl/secretbox"
 )
 
-func keyedHash(key, text []byte) []byte {
+func keyedHash(key *[hashLen]byte, text []byte) *[hashLen]byte {
 	h := keyedHasher(key)
 	h.Write([]byte(text))
-	return h.Sum(nil)
+	var sum [hashLen]byte
+	r := h.Sum(sum[:0])
+	if &r[0] != &sum[0] {
```

Comment:
I really like that we're checking the hash size. This line in particular clever, I assume it is based on how slices are implemented in Go or how the spec defines it. It's not immediately clear to me how this works though.

How are these addresses different if the actual Sum result size is unequal to the expected size? Does it work if the expected size is less than expected, as well as greater?

Will this work with gccgo? Not saying that's a hard requirement, but it is a possibility to consider.